File: barcodecore-mmo/mythiclib/src/main/java/io/lumine/mythic/lib/api/item/NBTItem.java
package io.lumine.mythic.lib.api.item;

import io.lumine.mythic.lib.MythicLib;
import io.lumine.utils.adventure.text.Component;

import org.bukkit.inventory.ItemStack;

import java.util.Arrays;
import java.util.List;
import java.util.Set;

public abstract class NBTItem {
    protected final ItemStack item;

    public NBTItem(ItemStack item) {
        this.item = item;
    }

    public ItemStack getItem() {
        return item;
    }

    public abstract Object get(String path);

    public abstract String getString(String path);

    public abstract boolean hasTag(String path);

    public abstract boolean getBoolean(String path);

    public abstract double getDouble(String path);

    public abstract int getInteger(String path);

    public abstract NBTCompound getNBTCompound(String path);

    public abstract NBTItem addTag(List<ItemTag> tags);

    public abstract NBTItem removeTag(String... paths);

    public abstract Set<String> getTags();

    public abstract ItemStack toItem();

    public abstract int getTypeId(String path);

    public abstract Component getDisplayNameComponent();

    public abstract void setDisplayNameComponent(Component component);

    public abstract List<Component> getLoreComponents();

    public abstract void setLoreComponents(List<Component> components);

    @Deprecated
    public abstract NBTItem cancelVanillaAttributeModifiers();

    public NBTItem addTag(ItemTag... tags) {
        return addTag(Arrays.asList(tags));
    }

    public double getStat(String stat) {
        return getDouble("MMOITEMS_" + stat);
    }

    public boolean hasType() {
        return hasTag("MMOITEMS_ITEM_TYPE");
    }

    public String getType() {
        String tag = getString("MMOITEMS_ITEM_TYPE");
        return !tag.equals("") ? tag : null;
    }

    public static NBTItem get(ItemStack item) {
        return MythicLib.plugin.getVersion().getWrapper().getNBTItem(item);
    }
}

File: barcode-framework/platform-bukkit-common/src/main/kotlin/com/vjh0107/barcode/framework/nbt/NBTItem.kt
package com.vjh0107.barcode.framework.nbt

import com.vjh0107.barcode.framework.koin.injector.inject
import com.vjh0107.barcode.framework.nbt.data.ItemTag
import com.vjh0107.barcode.framework.nms.NMSWrapper
import net.kyori.adventure.text.Component
import org.bukkit.inventory.ItemStack

abstract class NBTItem(val item: ItemStack) {
    /**
     * path 에 있는 NBTCompound 를 가져옵니다.
     */
    abstract fun getNBTCompound(path: String): NBTCompound

    /**
     * path 에 있는 값을 가져옵니다.
     */
    abstract operator fun get(path: String): Any?

    /**
     * path 에 있는 값을 String 으로 가져옵니다.
     */
    abstract fun getString(path: String): String

    /**
     * path 에 있는 값을 Boolean 으로 가져옵니다.
     */
    abstract fun getBoolean(path: String): Boolean

    /**
     * path 에 있는 값을 Double 으로 가져옵니다.
     */
    abstract fun getDouble(path: String): Double

    /**
     * path 에 있는 값을 Int 로 가져옵니다.
     */
    abstract fun getInt(path: String): Int

    /**
     * path 에 태그가 있는지 확인합니다.
     */
    abstract fun hasTag(path: String): Boolean

    /**
     * 태그를 추가합니다.
     */
    abstract fun addTag(tags: List<ItemTag>): NBTItem

    /**
     * 태그를 삭제합니다.
     */
    abstract fun removeTag(vararg paths: String): NBTItem

    /**
     * 태그 전체를 구합니다.
     */
    abstract val tags: Set<String>

    /**
     * Bukkit ItemStack 으로 변환합니다.
     */
    abstract fun toItem(): ItemStack

    /**
     * Type ID 를 구합니다.
     */
    abstract fun getTypeId(path: String): Int

    /**
     * DisplayName 을 구하거나 설정합니다.
     */
    abstract var displayNameComponent: Component

    /**
     * Lore 을 구하거나 설정합니다.
     */
    abstract var loreComponents: List<Component>

    /**
     * 태그를 추가합니다.
     */
    fun addTag(vararg tags: ItemTag): NBTItem {
        return addTag(listOf(*tags))
    }

    companion object {
        /**
         * NBTItem 을 가져옵니다.
         */
        fun get(item: ItemStack): NBTItem {
            val nmsWrapper: NMSWrapper by inject()
            return nmsWrapper.getNBTItem(item)
        }
    }
}

